File: 20190702/camouflage.test.js
// function getCloset(clothes) {
//   const closetObj = {};
//   clothes.forEach(v => closetObj[v[1]] === undefined ? closetObj[v[1]] = [v[0]] : closetObj[v[1]].push(v[0]));

//   return closetObj;
// }

// function getNumberOfCases(clothes) {
//   const closetObj = getCloset(clothes);
//   const typeOfClothes = Object.keys(closetObj);

//   let cases = 1;
//   typeOfClothes.forEach(v => cases *= (closetObj[v].length + 1));
//   cases--;

//   return cases;
// }


// test('getNumberOfCases', () => {
//   expect(getNumberOfCases([[`yellow_hat`, `headgear`], [`blue_sunglasses`, `eyewear`], [`green_turban`, `headgear`]])).toBe(5);
//   expect(getNumberOfCases([['crow_mask', 'face'], ['blue_sunglasses', 'face'], ['smoky_makeup', 'face']])).toBe(3);
// });

// test('getCloset', () => {
//   const closet = {
//     face: ['crow_mask', 'blue_sunglasses', 'smoky_makeup']
//   };
//   expect(getCloset([['crow_mask', 'face'], ['blue_sunglasses', 'face'], ['smoky_makeup', 'face']])).toMatchObject(closet);
// });

const obj = {
  i: 10,
  func() {
    console.log(this.i, this);
  }
}

console.log(obj.func());

File: 20200301/python/camouflage_test.py
from collections import Counter
from functools import reduce


def solution(clothes):
    c = Counter([category for _, category in clothes])

    return reduce(lambda acc, x: acc * (x + 1), c.values(), 1) - 1


def test_solution():
    assert solution(
        [["yellow_hat", "headgear"],
         ["blue_sunglasses", "eyewear"],
         ["green_turban", "headgear"]]) == 5
    assert solution(
        [["crow_mask", "face"],
         ["blue_sunglasses", "face"],
         ["smoky_makeup", "face"]]) == 3
